jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/Utf8Appendable.java

```diff
@@ -327,40 +331,159 @@ public boolean isUtf8SequenceComplete()
+     *
+     * <p>
+     *     If the input already has a replacement character, this is not counted when using
+     *     this method.
+     * </p>
+     *
+     * @return true if there was a replacement of an appended character or byte with a replacement character.
+     */
+    public boolean hasReplacements()
     {
-        if (!isUtf8SequenceComplete())
+        return _hasReplacements;
+    }
+
+    /**
+     * Get the UTF-8 encoded String.
+     *
+     * <p>
+     *     This will get the String as it exists currently, without including any trailing incomplete UTF-8 sequences.
+     * </p>
+     * <p>
+     *     Use {@link #finish()} to complete any incomplete UTF-8 sequences before calling this method.
+     * </p>
+     *
+     * @param throwOnReplacement to throw if a replacement to an appended character has occurred
+     * @return the UTF-8 String
+     * @throws NotUtf8Exception if String has replacement characters due to bad UTF-8 sequences
+     * @see #finish()
+     */
+    public String getString(boolean throwOnReplacement)
+    {
+        if (throwOnReplacement && hasReplacements())
         {
-            if (_coderErrorAction == CodingErrorAction.REPLACE)
-            {
-                try
-                {
-                    _appendable.append(REPLACEMENT);
-                }
-                catch (IOException e)
-                {
-                    throw new RuntimeException(e);
-                }
-                _hasReplacements = true;
-            }
-            _codep = 0;
-            _state = UTF8_ACCEPT;
-            if (_coderErrorAction == CodingErrorAction.REPORT)
-                throw new NotUtf8Exception("incomplete UTF8 sequence");
+            throw new NotUtf8Exception("String has invalid UTF-8 sequences");
+        }
+        return _appendable.toString();
+    }
+
+    /**
+     * Get the String as it exists currently.
+     *
+     * <p>
+     *     This will get the String as it exists currently, without including any trailing incomplete UTF-8 sequences.
+     * </p>
+     * <p>
+     *     Use {@link #finish()} to complete any incomplete UTF-8 sequences before calling this method.
+     * </p>
+     *
+     * @return the String as it exists currently, without throwing an Exception
+     * @see #finish()
+     */
+    public String getString()
+    {
+        return getString(false);
+    }
+
+    /**
+     * <p>
+     * Take the String from the {@link Appendable}, taking care to finish any incomplete UTF-8 sequences first.
+     * </p>
+     *
+     * <p>
+     * Calls to this method will {@link #reset()} this {@link Utf8Appendable}.
+     * </p>
+     *
+     * @return the String buffer from the appendable
+     * @throws CharacterCodingException if unable to encode the input bytes (eg: bad UTF-8 sequences)
+     */
+    @Override
+    public String takeString() throws CharacterCodingException
```

Comment:
Do we need to introduce CharacterCodingException?

Reply:
This is the signature of the `CharsetStringBuilder` interface, that Utf8Appendable always implemented.

Reply:
Hmmm i wonder if we can change our NotUtf8Exception to extend CharacterEncodingException and thus avoid the catch and rethrow... or is one checked?

Reply:
Dang it is checked. ignore.